runtime/compiler/runtime/J9Profiler.cpp

```diff
@@ -980,7 +980,7 @@ TR_ValueProfileInfoManager::getValueInfo(TR_ByteCodeInfo &bcInfo, TR::Compilatio
          info = _jitValueProfileInfo->getValueInfo(bcInfo, comp, kind, ArrayProfiler, true);
       }
 
-   if (external && (!info || info->getTotalFrequency() == 0))
+   if (external && (!info || info->getTotalFrequency() == 0) && kind == AddressInfo)
```

Comment:
This seems like a bit of a hack because IProfiler could supply any kind of data. Would it make more sense to make the kind available in some way (if it isn't already) and then test the kind at the point where we are expecting something that isn't an address so we don't use an address if we do get one?

Reply:
As far as I can tell, the IProfiler infrastructure doesn't currently provide any interface for figuring out the kind of data it's making available and only seems to be used for addresses at the moment. I'm not particularly familiar with the profiling code, so there might be something I missed there, though.

While I definitely agree that this is a bit of a hack and should be replaced by a better solution, this issue is currently blocking progress on eclipse/omr#4720 since it can cause the creation of invalid, though unreachable, BNDCHK nodes that later cause generation of invalid instructions, which is now a hard error under the new binary encoder. Would you be alright with having this temporary fix merged and an issue opened to document the need for a better solution to this?

Reply:
The IProfiler can return a number of different things - see the parse buffer method in https://github.com/eclipse/openj9/blob/982aecabceb311faeb432c21e2ab384ef1b052de/runtime/compiler/runtime/IProfiler.cpp#L3981. The IProfiler returns an index of a lookupswtich / tableswtich and returns bias information for branches.

The branch bias information is used for block frequency setting as is the lookupswitch / tableswitch. That being said the critical question is if anyone uses the value profiling API to get those. That I am not sure and I need to dig more. The answer to that question is critical to determining if this temporary fix can be merged as is. The risk here is performance regressions that are hard to isolate so I am hesitant to merge and fix it later unless we are 100% sure it isn't going to be break something...